### PR TITLE
[harfbuzz] port to new meson toolchain

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **harfbuzz**

Port harfbuzz to the new Meson toolchain, which is Harbuzz's primary build system.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by planetmarshall** [comment]
Bit of a mysterious one this - unable to reproduce and Microsoft docs are unhelpful.

```
freetyped.lib(ftdebug.obj) : fatal error LNK1318: Unexpected PDB error; OK (0) 'C:\J\w\prod\BuildSingleReference@2\s\9a2779\1\build-debug\util\freetype.pdb'
```

**Review by danimtb — APPROVED** [review]

File: recipes/harfbuzz/all/conanfile.py [hunk 1]
```diff
@@ -77,11 +80,11 @@ def validate(self):
             raise ConanInvalidConfiguration(
                 "Linking a shared library against static glib can cause unexpected behaviour."
             )
-        if scm.Version(self.version) >= "4.4.0":
-            if self.settings.compiler == "gcc" and scm.Version(self.settings.compiler.version) < "7":
+        if Version(self.version) >= "4.4.0":
+            if self.settings.compiler == "gcc" and Version(self.settings.compiler.version) < "7":
```
Comment (jwillikers) [inline recipes/harfbuzz/all/conanfile.py:84]: Is it possible to use the `self.settings.info` and `self.options.info` forms in the validate method here?

File: recipes/harfbuzz/all/conanfile.py [hunk 2]
```diff
@@ -92,48 +95,81 @@ def requirements(self):
...
+
+    def layout(self):
+        basic_layout(self, src_folder="src")
+
+    def generate(self):
+        def is_enabled(value):
+            return "enabled" if value else "disabled"
+
+        def meson_backend_and_flags():
+            if is_msvc(self) and Version(self.settings.compiler.version) == "15" and self.settings.build_type == "Debug":
```
Comment (jwillikers) [inline recipes/harfbuzz/all/conanfile.py:108]: This compiler version only corresponds to `Visual Studio` and not `msvc`. Can you add a check for `msvc`, too? We don't want this recipe to break for anyone using `msvc`.

File: recipes/harfbuzz/all/conanfile.py [hunk 3]
```diff
@@ -43,20 +58,8 @@ class HarfbuzzConan(ConanFile):
...
-        return "build_subfolder"
-
     def export_sources(self):
-        self.copy("CMakeLists.txt")
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
```
Comment (jwillikers) [inline recipes/harfbuzz/all/conanfile.py:65]: In the `configure` method, can you use `rm_safe` when deleting the `fPIC` option? You'll need to bump the minimum Conan version to 1.53, too.

File: recipes/harfbuzz/all/conanfile.py [hunk 4]
```diff
@@ -92,48 +95,81 @@ def requirements(self):
...
+            "freetype": is_enabled(self.options.with_freetype),
+            "gdi": is_enabled(self.options.get_safe("with_gdi")),
+            "directwrite": is_enabled(self.options.get_safe("with_directwrite")),
+            "gobject": is_enabled(can_run(self)),
+            "tests": "disabled",
+            "docs": "disabled",
+            "benchmark": "disabled",
+            "icu_builtin": "false"
+        })
+        tc.cpp_args += cxxflags
```
Comment (jwillikers) [inline recipes/harfbuzz/all/conanfile.py:129]: Is this variable documented for `MesonToolchain`?

File: recipes/harfbuzz/all/conanfile.py [hunk 5]
```diff
@@ -92,48 +95,81 @@ def requirements(self):
...
+                    full_folder = os.path.join(self.package_folder, libdir)
+                    for filepath in glob.glob(os.path.join(full_folder, f"*{ext}")):
+                        libname = os.path.basename(filepath)[0:-len(ext)]
+                        if remove_lib_prefix and libname[0:3] == "lib":
+                            libname = libname[3:]
+                        rename(self, filepath, os.path.join(os.path.dirname(filepath), f"{libname}.lib"))
+
+        meson = Meson(self)
+        meson.install()
+        fix_msvc_libname()
```
Comment (jwillikers) [inline recipes/harfbuzz/all/conanfile.py:168]: Should [fix_apple_shared_install_name](https://docs.conan.io/en/latest/reference/conanfile/tools/apple.html#conan-tools-apple-fix-apple-shared-install-name) also be used here?

File: recipes/harfbuzz/all/test_package/conanfile.py [hunk 6]
```diff
@@ -1,19 +1,28 @@
...
 
     def test(self):
-        if not tools.cross_building(self):
+        if can_run(self):
             font = os.path.join(self.source_folder, "example.ttf")
-            bin_path = os.path.join("bin", "test_package")
-            self.run("{} {}".format(bin_path, font), run_environment=True)
-
+            bin_path = os.path.join(self.cpp.build.bindirs[0], "test_package")
+            self.run("{} {}".format(bin_path, font), env="conanrun")
```
Comment (jwillikers) [inline recipes/harfbuzz/all/test_package/conanfile.py:28]: Can this be a Python F-string?

File: recipes/harfbuzz/all/test_v1_package/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self):
+            font = os.path.join(self.source_folder, "..", "test_package", "example.ttf")
+            bin_path = os.path.join("bin", "test_package")
+            self.run("{} {}".format(bin_path, font), run_environment=True)
```
Comment (jwillikers) [inline recipes/harfbuzz/all/test_v1_package/conanfile.py:18]: Python F-string?

**Review by jwillikers — CHANGES_REQUESTED** [review]

File: recipes/harfbuzz/all/conandata.yml [hunk 8]
```diff
@@ -18,21 +18,13 @@ sources:
...
     sha256: "77c8c903f4539b050a6d3a5be79705c7ccf7b1cb66d68152a651486e261edbd2"
 patches:
-  "3.2.0":
-    - patch_file: "patches/icu-3.0.x.patch"
-      base_path: "source_subfolder"
   "4.4.1":
-    - patch_file: "patches/icu-4.0.x.patch"
-      base_path: "source_subfolder"
+    - patch_file: "patches/0000-fix-freetype-lookup-4.4.1.patch"
+      patch_type: "backport"
```
Comment (prince-chrismc) [inline recipes/harfbuzz/all/conandata.yml:23]:
```suggestion
      patch_type: "portability"
```

File: recipes/harfbuzz/all/conandata.yml [hunk 9]
```diff
@@ -18,21 +18,13 @@ sources:
...
-    - patch_file: "patches/icu-5.2.x.patch"
-      base_path: "source_subfolder"
-  "5.3.0":
-    - patch_file: "patches/icu-5.2.x.patch"
-      base_path: "source_subfolder"
-  "5.3.1":
-    - patch_file: "patches/icu-5.2.x.patch"
-      base_path: "source_subfolder"
+    - patch_file: "patches/0000-fix-freetype-lookup-5.1.0.patch"
+      patch_type: "backport"
```
Comment (prince-chrismc) [inline recipes/harfbuzz/all/conandata.yml:28]:
```suggestion
      patch_type: "portability"
```

**Review by prince-chrismc — COMMENTED** [review]
Backport was a misleading name... https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#backport is for changes to source code

This patch better matches portability :)

Trying to make these consistent 🙏

File: recipes/harfbuzz/all/conanfile.py (hunk 4 above)
Comment (planetmarshall) [reply recipes/harfbuzz/all/conanfile.py:129]: Yes, although the documentation isn't all that clear for the `cpp_*` attributes.

Screenshot from 2022-11-11 16-33-30

Comment (jwillikers) [reply recipes/harfbuzz/all/conanfile.py:133]: Hmm, I don't see any recipes using this yet. @uilianries Is it safe to use `cpp_args` in Meson toolchain? It seems like the best choice here, but is only mentioned in a bit of an off-hand way in the documentation.

**Review by AndreyMlashkin — APPROVED** [review]

**Comment by AndreyMlashkin** [comment]
awesome contribution, thanks @planetmarshall

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
This is a great lift and looks really good.

If there's small issues id sugggest making a followup PR so we can get this out

**Pull request merged**